docs/src/lib/metadata.md

```diff
@@ -149,24 +150,25 @@ data frames.
     Such changes might be made based on users' feedback about what metadata
     propagation rules are most convenient in practice.
 
-The general design rules for propagation of table and column level metadata
-are as follows:
-* if some object (data frame or column) is mutated in-place its metadata
-  is not changed.
-* preferably (this is not guaranteed in all cases as sometimes it might be hard
-  to detect that this condition is met) if some object is moved as a whole
-  (either by coping or by aliasing) then its metadata is propagated.
-* in all other cases if a new object is allocated from an existing object and
-  this transformation is not a copy or aliasing then metadata is dropped.
+The general design rules for propagation of table metadata is as follows:
+* for all operations that take a single data frame like object
+  and return a data frame like object table level metadata is preserved;
+* for all operations that take more than one data frame like object
+  and return a data frame like object (e.g. `hcat` or joins) table level metadata
+  is preserved from all tables with the exception that if two or more tables
+  define the same metadata then it is silently dropped.
 
-The rationale behind these rules is that metadata can be arbitrary information
-about the object and any change of the object that the metadata is attached to
-might invalidate it.
+The general design rules for propagation of column metadata is as follows:
+* if the column values are not changed in the operation and column name is not
+  changed then column metadata is retained;
+* when renaming columns using `rename!` and `rename` column metadata is
+  retained (it is considered to be attached to column value).
```

Comment:
Maybe we can propagate metadata even if the column name changes (e.g. `select(df, :x => :y)`)? There's no reason to think that it wouldn't apply anymore. That would avoid the need of a special case for `rename`.

Reply:
It will have O(number of columns^2) cost then, which I want to avoid. If we lift the restriction (except for `rename` and `rename!` - in which case we can cheaply propagate metadata) I would have to check all input columns against all output columns for metadata. With the restriction the cost is O(number of columns).

Reply:
However, I have added a rule that when automatic column renaming happens (can happen in joins and `hcat`) then we retain metadata.

Reply:
I was just thinking about special casing `:x => :y` (and maybe `:x => identity => :y` if that makes the implementation easier). That sounds natural.

Reply:
OK - adding support for both (they are indistinguishable internally) should be doable but it will be a special work-around both in `DataFrame` and `GroupedDataFrame`.